cli/formatters.py

```diff
@@ -5,14 +5,17 @@
 from enum import Enum
 from functools import partial
 from json import JSONEncoder
-from typing import Any, Callable, Optional
+from typing import Any, Callable, Optional, Protocol
 
 from tabulate import tabulate
 
 from cli import errors
 from cli.utils import flatten
 
-Formatter = Callable[[Any], str]
+
+class Formatter(Protocol):
```

Comment:
You need to inherit from this in each of the formatters so that `get_formatter` types are right.

replace the `do_fmt` func with 

```python
    def do_fmt(self, value: Any) -> str:
        return fmt_any(value, fmt_list=self.fmt_list)
```
then add  `fmt_list` and `to_dict` and a class const for `desired_fmt` that each class overwrite
```python
    def to_dict(x):
        return to_dict_or_raise(x,desired_fmt=desired_fmt)
    def fmt_list(self, xs: list) -> str:
        raise NotImplementedError("Missing implemenation of formater list")
```
Then you can ovrewrite `do_fmt` at `JsonFormatter`, and `fmt_list` at the rest

It might be a bigger hustle but `JsonFormatter` could be simplified even further if we extract the logic there to use these functions, like "to_dict" for example, but I realize this may be too much for this PR.